soc: nordic: nrf53: assign pin xl1,xl2 to app core if lfxo disabled

ivynya opened the pull request:
I opened an issue #92663 where disabling LFXO for the nRF53 via devicetree or kconfig does not properly allow the pins from the LFXO (0.00 and 0.01) to be used as GPIO. This adds a few guards against those devicetree and kconfig properties, and if LFXO is disabled by the build configuration, will assign those pins to the nRF53 app core instead.

bjarki-andreasen reviewed (changes requested):
I believe a better option than riding on the LFXO_CAP, would be to just use `DT_NODE_HAS_STATUS_OKAY(LFXO_NODE)`, and if it is, then `BUILD_ASSERT(DT_NODE_HAS_PROP(LFXO_NODE), load_capacitors)` and lastly, clean up the`#ifdef`  like
```
#ifdef CONFIG_SOC_NRF5340_CPUAPP
#if DT_NODE_HAS_STATUS_OKAY(LFXO_NODE)
	nrf_oscillators_lfxo_cap_set(NRF_OSCILLATORS, LFXO_CAP);
#if !defined(CONFIG_BUILD_WITH_TFM)
	/* This can only be done from secure code.
	 * This is handled by the TF-M platform so we skip it when TF-M is
	 * enabled.
	 */
	nrf_gpio_pin_control_select(PIN_XL1, NRF_GPIO_PIN_SEL_PERIPHERAL);
	nrf_gpio_pin_control_select(PIN_XL2, NRF_GPIO_PIN_SEL_PERIPHERAL);
#endif /* !defined(CONFIG_BUILD_WITH_TFM) */
#elif !defined(CONFIG_BUILD_WITH_TFM)
	nrf_gpio_pin_control_select(PIN_XL1, NRF_GPIO_PIN_SEL_APP);
	nrf_gpio_pin_control_select(PIN_XL2, NRF_GPIO_PIN_SEL_APP);
#endif
```
This aligns everything to use the devicetree, and is is clearer as `DT_NODE_HAS_STATUS_OKAY(LFXO_NODE)` means what it says, where LFXO_CAP is being used to mean the same, but IMO does not read the same :)

If maintaining the HWM1 module is of interest, then a bespoke PR for it based on the deprecated Kconfig symbol would be needed.

ivynya commented:
Thank you for the feedback! I believe I applied the suggestions correctly, checking directly and exclusively for the devicetree node instead. I also removed the legacy LFXO Kconfig if block (L62-72) as it sounds like after these changes it will not affect anything anymore. I tested quickly with my project and it seems to work. Please let me know if I might have misinterpreted anything or further changes are needed.

bjarki-andreasen reviewed (dismissed):
This looks just about perfect :) Nice work!

bjarki-andreasen commented:
The Signed-off-by: entry must use the full name of your github profile, so `Signed-off-by: Ivynya Lu <ivy.lu@level.co>`

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -54,22 +54,14 @@
 #define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_EXTERNAL
 #elif DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, internal)
 #define LFXO_CAP (DT_ENUM_IDX(LFXO_NODE, load_capacitance_picofarad) + 1U)
-#else
-/* LFXO config from legacy Kconfig */
-#if defined(CONFIG_SOC_LFXO_CAP_INT_6PF)
-#define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_6PF
-#elif defined(CONFIG_SOC_LFXO_CAP_INT_7PF)
-#define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_7PF
-#elif defined(CONFIG_SOC_LFXO_CAP_INT_9PF)
-#define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_9PF
-#else
-#define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_EXTERNAL
-#endif
-#endif
```
This part cannot be removed yet. These options are deprecated but they are still present in Kconfig, so they still need to be supported.

anangl replied:
Yes, they were deprecated in 4.0, so the can be removed now.

anangl commented on soc/nordic/nrf53/soc.c:
```diff
@@ -54,22 +54,14 @@
 
 /* LFXO config from DT */
+#if DT_NODE_HAS_STATUS_OKAY(LFXO_NODE)
+BUILD_ASSERT(DT_NODE_HAS_PROP(LFXO_NODE, load_capacitors));
```
I don't think I understand the issue. Properties with `required: true` are required by dtc only for nodes with status "okay", so for disabled nodes you don't need to provide any value for them.

bjarki-andreasen reviewed (changes requested):
Looking into the build failure, some boards depend on the previous default behavior, which automatically sets load capacitors to "external" if `load-capacitors` is not specified in the devicetree.

To avoid having to look through and update all boards to check if they set load-capacitors to manually set them to external, let's keep the default behavior. That is, don't assert
```
BUILD_ASSERT(DT_NODE_HAS_PROP(LFXO_NODE, load_capacitors))
```
and instead default to external if the prop does not exist:
```
#if DT_NODE_HAS_PROP(LFXO_NODE, load_capacitors)
#if DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, external)
#define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_EXTERNAL
#elif DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, internal)
#define LFXO_CAP (DT_ENUM_IDX(LFXO_NODE, load_capacitance_picofarad) + 1U)
#endif
#else
#define LFXO_CAP NRF_OSCILLATORS_LFXO_CAP_EXTERNAL
#endif
```
and leave the binding, so load-capacitors stays not required :)

bjarki-andreasen reviewed (dismissed):
Looks really good!

ivynya committed:
soc: nordic: nrf53: assign pin xl1,xl2 to app core if lfxo disabled

Fixes zephyrproject-rtos/zephyr#92663

Disabled LFXO via devicetree allows pin 0.00 and 0.01 to work correctly
as gpio by assigning it to the app core instead of peripheral. Removed
deprecated Kconfig options so DT is the only config path now.

Signed-off-by: Ivynya Lu <ivy.lu@level.co>

bjarki-andreasen commented:
ping @anangl

KajbaM commented:
Hello, first of all thank you for providing this, after tracking this problem, this PR was very helpful for providing patch to the project I am currently working on.

I would however like to point out, that this patch on its own only solves the issue for secure build. Same problem this resolves is still present when using `/ns` builds due to same bug being present in TF-M image. I am unfortunately unable to open issue regarding this in trusted firmware repository as issues page doesn't seem to exist there.

Once this gets merged similar change should be done to trusted-firmware-m repository.
